sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,184 @@
+        // TODO: glog timeout
+    } else {
+        // TODO: glog no timeout
+    }
+}
+
+void BatchSpanProcessor::DoBackgroundWork(){
+    int timeout = schedule_delay_millis_;
+
+    while (true)
+    {
+        // We wait for the buffer to fill up to max_export_batch_size 
+        //                     OR
+        // We see if the force_flush() method has been invoked
+        if(static_cast<int>(buffer_->size()) < max_export_batch_size_ && 
+           is_force_flush_ == false)
+           {
+            std::unique_lock<std::mutex> lk(cv_m_);
+            cv_.wait_for(lk, std::chrono::milliseconds(timeout));
+
+            // If shutdown has been invoked, break out of the loop
+            // and drain out the queue.
+            if(is_shutdown_) break;
+
+            // In case of spurious wake up, we export only if we have atleast one span in the batch.
+            // This is acceptable because batching is a best mechanism effort here.
+            if(static_cast<int>(buffer_->size()) == 0) continue;
+        }
+
+        auto start = std::chrono::steady_clock::now();
+        Export();
+        auto end = std::chrono::steady_clock::now(); 
+        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+        timeout = schedule_delay_millis_ - duration.count();
+    }
+
+    // Drain the queue in case there are any spans still left. 
+    // Invoked when shutdown() is called.
+    DrainQueue();
+}
+
+void BatchSpanProcessor::Export(){
+    // Get the appropriate size
+    const int buffer_size = static_cast<int>(buffer_->size());
+    const int num_spans_to_export = buffer_size >= max_export_batch_size_ ? max_export_batch_size_ : buffer_size;
+
+    std::vector<std::unique_ptr<Recordable>> spans_arr;
+    
+    bool is_force_flush_token_present = false;
+
+    buffer_->Consume(
+    num_spans_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+        std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+        ptr.Swap(swap_ptr);
+
+        // Check if we just processed/consumed the force flush token
+        if(swap_ptr == nullptr) is_force_flush_token_present = true;
+
+        else spans_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+        return true;
+        });
+    });
+
+    exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));
+
+    // If we processed the force flush token, notify
+    if (is_force_flush_token_present) 
+    {
+        force_flush_cv_.notify_one();
+    }
+    
+}
+
+void BatchSpanProcessor::DrainQueue(){
+    while(static_cast<int>(buffer_->size())) Export();
+}
+
+void BatchSpanProcessor::Shutdown(std::chrono::microseconds timeout) noexcept 
+{
+    is_shutdown_ = true;
+
+    // Notify thread, drain the queue and wait for the thread
+    cv_.notify_one();
```

Comment:
It's very likely that we miss the notification call here too. Probably have to come up with a way to forcefully kill/interrupt a thread.
Any thoughts on this, @reyang?

Reply:
I think you can loop on notification using is_shutdown_ as a marker to ensure it gets notified:

std::chrono::microseconds timeout_left = timeout;
while(is_shudown_ && timeout_left >0) {
auto start = std::chrono::steady_clock::now();
cv_.notify_one();
std::this_thread::sleep_for(std::chrono::nanoseconds(50));
auto end = std::chrono::steady_clock::now();
timeout_left -= (end-start);
}
Setting is_shutdown_ to false in doBackGroundWork if it  was set to true when it wakes up.